lib/net.js

```diff
@@ -631,10 +631,20 @@ Socket.prototype.destroySoon = function() {
   if (this.writable)
     this.end();
 
-  if (this.writableFinished)
+  if (this.writableFinished && this.readableEnded)
     this.destroy();
   else
-    this.once('finish', this.destroy);
+    this.once('finish', () => {
+      // If we destroy the socket before we have received an incoming EOF
+      // the remote might still be sending data that will trigger a RST
+      // from this host once it is received
+      // Closing a TCP socket requires two FIN/ACK exchanges, one in every
+      // direction - this means two events
+      if (this.readableEnded)
+        this.destroy();
+      else
+        this.once('end', this.destroy);
+    });
```

Comment:
Doesn't `autoDestroy` already to the exact same thing?

Reply:
I played a lot with it, I was not able to get it to correctly close the connection
But now we understand the connection close mechanism: `uv_shutdown()` is to be called, in its callback `finish` is to be emitted, and then reading must continue until `end`
Maybe this can be reimplemented in terms of `autoDestroy`
Everything that is needed is a way to trigger this from `res.end()`

Reply:
Or it could happen the other way around - you get an `end`, you call the shutdown, then when the callback is executed, you emit `finish`!